### PR TITLE
Implementation of NARGP

### PR DESCRIPTION
**Pull request by jibanCat**
Use the toy functions from https://github.com/EmuKit/emukit/blob/80def9249a7a42854fe91a1f336ad7c18c1a572a/notebooks/Emukit-tutorial-multi-fidelity.ipynb

I still need to refine the notebooks and add comments. But all the information are there.

And unfortunately I don't think this implementation is faster than GPy implementation.

Due to the property of the noiseless toy examples, I have to fix the observation noise to very small.

A few parts I am not sure:
* During training, do I do Monte Carlo sample from the Low-Fidelity node are do the mean prediction?
* During predicting, do I do Monte Carlo sample from the Low-Fidelity node are do the mean prediction?
* During training, can I optimize LF and HF together? (I am but GPy implementation separates the training.)
* During training, should I specify prior for LF kernel in NARGP kernel? (I am. I tested there's no difference with/without specifying LF kernel hyperparameter prior.)

### CODE REVIEW

**Comment by jibanCat** [comment]
Try a bit reproducing my previous MFEmulator and MFBox papers using TFP.

The optimization is a bit tricky I ended up using Uniform priors (the NARGP kernel several hyperparameters are degenerate, if imposing a peaked prior, then it goes to non-optimal likelihood region), which not very Bayesian.


TODO 

* Might need to play around a bit to find which hyperparameter is more sensitive to prior choice.

* Could also just run a HMC to see the posteriors

**Comment by jibanCat** [comment]
MFBox test spectrum:

image


Training spectra (low-fidelity already reduced the mean to remove the mean particle spacing bias. But in practice it works without remove the mean, the results changed within 2% at the smallest scales.)

image

Training history (using Adam optimizer. First: Training L1 and L2 with noise fixed. Second, noise relaxed. Third, Training MFBox with L1 and L2 fixed, and noise fixed. Final, Training MFBox but noise relaxed, L1 and L2 fixed.)

image

Pred / Exact (One test spectrum looks weird, might be because the uniform priors make the optimization very degenerate)

image

**Comment by jibanCat** [comment]
NARGP example, reproducing the one on adapt website: https://adapt-python.github.io/adapt/examples/Multi_fidelity.html

On the website: 


The current NARGP in the branch (notebook 13):

image

Note that I did a normalization of Y here to make sure the mean function close to zero, which might make the extrapolation part close to the true function. Note that this is a clever choice, not necessary always match at the extrapolation region.

When no normalization of Y, it will regress to the mean:

image

**Comment by jibanCat** [comment]
> TODO
> 
> * Might need to play around a bit to find which hyperparameter is more sensitive to prior choice.

The interaction kernel is sensitive to the prior. The length scale is super large with very small amplitude. Basically suggest $\rho$ has weak parameter dependent.

Observation noise needs to start with a small fixed value. Otherwise the variance will just take over the likelihood making the GP prediction with super wide uncertainty.

> * Could also just run a HMC to see the posteriors

Might not worth it. The dynamic range of the hyperparameters is large (I meant the different peaks in the posterior distribution the difference is large).

**Comment by jibanCat** [comment]
TODO: Clean up and better documentation (for at least one NARGP notebook)
